service/history/chasm_engine.go

```diff
@@ -241,14 +256,137 @@ func (e *ChasmEngine) ReadComponent(
+		return satisfiedRef, nil
+	}
+
+	// Wait condition not satisfied; long-poll
+
+	namespaceRegistry, err := shardContext.GetNamespaceRegistry().GetNamespaceByID(
+		namespace.ID(requestRef.EntityKey.NamespaceID),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	internalLongPollTimeout := shardContext.GetConfig().LongPollExpirationInterval(namespaceRegistry.Name().String())
+	ctx, cancel := context.WithTimeout(ctx, internalLongPollTimeout)
+	defer cancel()
+
+	// PollComponent subscribes to execution-level notifications. Suppose that an execution consists
+	// of one component A, and A has subcomponent B. Subscribers interested only in component B may
+	// be woken up unnecessarily (and thus evaluate the predicate unnecessarily) due to changes in
+	// parts of A that do not also belong to B.
+	ch := e.notifier.Subscribe(requestRef.EntityKey)
+	executionLease.GetReleaseFn()(nil)
+
+	for {
+		select {
+		// It is possible that multiple state transitions (multiple notifications) occur between
+		// predicate checks, therefore the predicate must be monotonic: if it returns true at
+		// execution state transition s then it must return true at all transitions t > s.
+		case <-ch:
+			_, executionLease, err := e.getExecutionLease(ctx, requestRef)
+			if err != nil {
+				return nil, err
+			}
+			func() {
+				defer executionLease.GetReleaseFn()(nil)
+				satisfiedRef, err = e.predicateSatisfied(ctx, monotonicPredicate, requestRef, executionLease)
+				if err == nil && satisfiedRef == nil {
+					ch = e.notifier.Subscribe(requestRef.EntityKey)
+				}
+			}()
+			if err != nil {
+				return nil, err
+			}
+			if satisfiedRef != nil {
+				return satisfiedRef, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+// predicateSatisfied is a helper function for PollComponent. It returns (ref, err) where ref is non-nil
+// iff there's no error and predicate evaluates to true.
+func (e *ChasmEngine) predicateSatisfied(
+	ctx context.Context,
+	predicate func(chasm.Context, chasm.Component) (bool, error),
+	ref chasm.ComponentRef,
+	executionLease api.WorkflowLease,
+) ([]byte, error) {
+	chasmTree, ok := executionLease.GetMutableState().ChasmTree().(*chasm.Node)
+	if !ok {
+		return nil, serviceerror.NewInternalf(
+			"CHASM tree implementation not properly wired up, encountered type: %T, expected type: %T",
+			executionLease.GetMutableState().ChasmTree(),
+			&chasm.Node{},
+		)
+	}
+
+	// It is not acceptable to declare the predicate to be satisfied against execution state that is
+	// behind the requested reference. However, getExecutionLease does not currently guarantee that
+	// execution VT >= ref VT, therefore we call IsStale() again here and return any error (which at
+	// this point must be ErrStaleState; ErrStaleReference has already been eliminated).
+	err := chasmTree.IsStale(ref)
+	if err != nil {
+		// ErrStaleState
+		// TODO(dan): this should be retryable if it is the failover version that is stale
```

Comment:
Shouldn't it always be retryable? Seems like we should return unavailable errors in this case assuming that we've refreshed the state.